### PR TITLE
Added section on C# Source Generators

### PR DESCRIPTION
**Pull request by hermanussen**
C# Source Generators are a new .NET 5 Roslyn feature that is currently in preview. Some really interesting ideas are popping up, and I think it would be great if they were listed here.

Strictly speaking, most of these projects have not gathered enough popularity to be featured in this list. But because of the "early-stage" nature of this, I'm hoping this can be overlooked for now. It can always be curated later.

If you have more source generators that seem interesting, please add them here.

### CODE REVIEW

**Comment by amis92** [comment]
I've started a separate list (not awesome since they're not yet provably-awesome ;) ):

https://github.com/amis92/csharp-source-generators

I'd love to add anything you might know about.

**Comment by ironcev** [comment]
@hermanussen, @amis92 thanks a lot for your contributions, and please forgive me for ignoring this PR for so long! Professional and private commitments are hitting me hard at the moment and I wanted to give this a proper thought like for all other entries in the Awesome Roslyn list.

My (initial) five cents:

- Generators are a mighty feature I was looking forward to and wishing it from the first moment I saw a proposal draft. I expected an explosion of generators once the feature is out and your research shows that exactly that happened. I am definitely for having a separate section on generators as proposed by @hermanussen.
- Early-nature and maturity is not an issue if the quality and potential are there. As stated in the [contributing.md](https://github.com/ironcev/awesome-roslyn/blob/master/contributing.md#open-source-projects): "*Note that some awesome projects do not fulfill the above maturity criteria. For example, Microsoft.CodeAnalysis.CSharp.PatternMatching provides a groundbreaking and inspiring approach to pattern matching on Roslyn syntax trees. As such it fully deserves to be on the list, although the author clearly stated that he does not plan to maintain it actively.*"
- @amis92 your list *is* awesome! Thank you for all the research!

I would propose the following. Please give me a day or two to get a close look at the links you've provided. Then we can shortly brainstorm how to integrate them. E.g. should the articles go to the general tutorial section, should we take the [csharp-source-generators](https://github.com/amis92/csharp-source-generators) over, all of them or some or reference to the list, or.... I am open for all suggestions and actually overwhelmed with the amount and quality of your contributions! So far it was mostly single entries, this one is really huge :-)

A big thanks once again and please give me a day or two to draft a proposal on how to combine all this together. Also please give your thoughts in the meanwhile.

Thanks!

**Comment by ironcev** [comment]
Ok, I realized I will not be able to sleep tonight if don't sketch at least the very first proposal :-)

What do you think of this?

The Tutorials section gets divided into two subsections:

- Tutorials
  - Analyzers, code fixes, and refactorings
  - Source Generators

And we add an additional top-level section, named so that it fits the rest of the naming scheme:

- Open Source C# 9.0 Source Generators

Articles and projects that you've found can go to these new sections.  The videos can go to the existing Talks section.

In addition, I would take a closer look at the projects and articles and create a selection (maybe all will fit, why not!). The idea of an awesome list is to have a selection, and not a collection. I see a benefit of the more collection-like approach like https://github.com/amis92/csharp-source-generators but I wouldn't mix the two?

How does this sound to you?

@amis92 How do you see the future of the csharp-source-generators list? Should it stay in your opinion as a collection list and serve as a source for the awesome list? If so, we should link it from the awesome list at least. Or you think that having Source Generators on the awesome list serves the purpose? As mentioned, I see the benefits of both approaches.

A big thanks once again and let me know what two of you think.

Looking forward to extending the list with this new content and exciting new language feature!

**Comment by amis92** [comment]
Well, I think it's quite tangled in that this whole brave new SourceGen world is *in general* awesome ;) But you can't really say which projects are awesome right now. Which is why I've specifically not named my repo `awesome-...` ;D

The topic is huge, so the more resources the better. I'll definitely try to keep my list as inclusive as possible, as I feel it's really useful especially in early stages to be able to look at other projects and get a feeling of how to do that.

> @amis92 your list is awesome! Thank you for all the research!
Well I'd be humbled to have my list included here then ;)

TLDR: I think a category dedicated to Source Generators here is very much deserved, but I don't think it'd serve the purpose my list attempts to (at least for some time), so I'll continue working on it as well.

**Comment by ironcev** [comment]
Perfect, let's do it like that! I'll accept this PR (thanks Robin!), rearrange entries a bit, see what to include from the csharp-code-generators, and link to that repo. I'll get to it most likely by the end of the week. Looking forward :-)
